crates/evm/fuzz/src/strategies/invariants.rs

```diff
@@ -168,17 +180,18 @@ pub fn fuzz_contract_with_calldata(
     fuzz_fixtures: &FuzzFixtures,
     contract: Address,
     func: Function,
-) -> impl Strategy<Value = (Address, Bytes)> {
+    contract_abi: JsonAbi,
+) -> impl Strategy<Value = CallDetails> {
     // We need to compose all the strategies generated for each parameter in all possible
     // combinations.
     // `prop_oneof!` / `TupleUnion` `Arc`s for cheap cloning.
     #[allow(clippy::arc_with_non_send_sync)]
     prop_oneof![
         60 => fuzz_calldata(func.clone(), fuzz_fixtures),
-        40 => fuzz_calldata_from_state(func, fuzz_state),
+        40 => fuzz_calldata_from_state(func.clone(), fuzz_state),
     ]
     .prop_map(move |calldata| {
         trace!(input=?calldata);
-        (contract, calldata)
+        CallDetails::new(contract, calldata, func.clone(), contract_abi.clone())
```

Comment:
another approach here would be to not include function and target abi in `CallDetails` but instead lookup the abi / identify function (using calldata) from `ContractsByAddress` when collecting values from call. I find including them in call details nicer but happy to change if you see any side effect